### PR TITLE
Add atreides03a

### PR DESCRIPTION
**Pull request by abcdefg30**

### CODE REVIEW

**Comment by MustaphaTR** [comment]
Engineer comes at 4th levels. It supposed to be unavalibe at there. I think Medic too.

At orginal game Upgrade hint comes right after you build a Barracks or Light factory. Here it comes 32 seconds later than it.

File: mods/d2k/maps/atreides-03a/map.yaml [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			Prerequisites: ~disabled
+	heavy_factory:
+		Buildable:
+			Prerequisites: ~disabled
+	medium_gun_turret:
+		Buildable:
+			Prerequisites: ~disabled
+	wall:
+		Buildable:
+			Prerequisites: ~disabled
```
Comment (MustaphaTR) [inline mods/d2k/maps/atreides-03a/map.yaml]: Walls are avalible at 3. Missions.

Comment (Biofreak1987) [inline mods/d2k/maps/atreides-03a/map.yaml]: This tile at the start doesn't look right:

tile_error.

**Comment by abcdefg30** [comment]
> Engineer comes at 4th levels. It supposed to be unavalibe at there. I think Medic too.

Derp, fixed. I don't know why I enabled them in the first place...

> At orginal game Upgrade hint comes right after you build a Barracks or Light factory. Here it comes 32 seconds later than it.

That just depends on when you build the facilities.
(See the `GameTime % 32seconds` check.)

**Comment by abcdefg30** [comment]
Updated.

**Comment by Biofreak1987** [comment]
This tile at the start doesn't look right: (Posted that in the wrong place before)

5b5ef1a8-7586-11e5-9469-869f6f515a73

**Comment by abcdefg30** [comment]
> This tile at the start doesn't look right: (Posted that in the wrong place before)

Should be already fixed.

**Comment by Biofreak1987** [comment]
Ah ok then, I just moved that post without rechecking the update. Sorry about that.

**Comment by abcdefg30** [comment]
No problem.

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

File: mods/d2k/missions.yaml [hunk 2]
```diff
@@ -2,4 +2,4 @@ Atreides Campaign:
 	./mods/d2k/maps/atreides-01a
 	./mods/d2k/maps/atreides-01b
 	./mods/d2k/maps/atreides-02a
-	./mods/d2k/maps/atreides-02b
+	./mods/d2k/maps/atreides-02b	./mods/d2k/maps/atreides-03a
```
Comment (RoosterDragon) [inline mods/d2k/missions.yaml]: Missing newline?

Comment (abcdefg30) [inline mods/d2k/missions.yaml]: Ops, seems like I screwed this up while rebasing.

**Comment by abcdefg30** [comment]
Fixed.

File: mods/d2k/maps/atreides-03a/atreides03a-AI.lua [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+SetupAttackGroup = function()
+	local units = { }
+
+	for i = 0, AttackGroupSize[Map.Difficulty], 1 do
+		if #IdlingUnits == 0 then
+			return units
+		end
+
+		local number = Utils.RandomInteger(1, #IdlingUnits)
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/atreides03a-AI.lua]: `#IdlingUnits + 1`, our `RandomInteger` excludes the upper bound.

File: mods/d2k/maps/atreides-03a/atreides03a-AI.lua [hunk 4]
```diff
@@ -0,0 +1,164 @@
+
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/atreides03a-AI.lua]: spurious newline

File: mods/d2k/maps/atreides-03a/atreides03a-AI.lua [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+	if OBarracks.IsDead then
+		return
+	end
+
+	if HoldProduction then
+		Trigger.AfterDelay(DateTime.Minutes(1), ProduceInfantry)
+		return
+	end
+
+	local delay = Utils.RandomInteger(AttackDelays[Map.Difficulty][1], AttackDelays[Map.Difficulty][2])
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/atreides03a-AI.lua]: here too: `AttackDelays[Map.Difficulty][2] + 1`

File: mods/d2k/maps/atreides-03a/atreides03a-AI.lua [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	if OLightFactory.IsDead then
+		return
+	end
+
+	if HoldProduction then
+		Trigger.AfterDelay(DateTime.Minutes(1), ProduceVehicles)
+		return
+	end
+
+	local delay = Utils.RandomInteger(AttackDelays[Map.Difficulty][1], AttackDelays[Map.Difficulty][2])
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/atreides03a-AI.lua]: and here

File: mods/d2k/maps/atreides-03a/atreides03a.lua [hunk 7]
```diff
@@ -0,0 +1,175 @@
+
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/atreides03a.lua]: spurious newline

File: mods/d2k/maps/atreides-03a/atreides03a.lua [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	return #player.GetActorsByType(AtreidesBaseBuildings[index]) > 0 and not player.HasPrerequisites({ AtreidesUpgrades[index] })
+end
+
+Tick = function()
+	if player.HasNoRequiredUnits() then
+		ordos.MarkCompletedObjective(KillAtreides)
+	end
+
+	if ordos.HasNoRequiredUnits() and not player.IsObjectiveCompleted(KillOrdos) then
+		Media.DisplayMessage("The Ordos have been anihilated!", "Mentat")
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/atreides03a.lua]: "annihilated".

File: mods/d2k/maps/atreides-03a/atreides03a.lua [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+	if player.Resources > ToHarvest[Map.Difficulty] - 1 then
+		player.MarkCompletedObjective(GatherSpice)
+	end
+
+	if DateTime.GameTime % DateTime.Seconds(32) == 0 and (MessageCheck(1) or MessageCheck(2)) then
+		Media.DisplayMessage("Upgrade barracks and light factory to produce more advanced units.", "Mentat")
+	end
+
+	if HasPower and player.Resources > player.ResourceCapacity * 0.8 and DateTime.GameTime % DateTime.Seconds(32) == 0 then
+		Media.DisplayMessage(Messages[4], "Mentat")
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/atreides03a.lua]: Is this a left-over? I can't find a `Messages` array anywhere.

File: mods/d2k/maps/atreides-03a/atreides03a.lua [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+end
+
+InitObjectives = function()
+	Trigger.OnObjectiveAdded(player, function(p, id)
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "New " .. string.lower(p.GetObjectiveType(id)) .. " objective")
+	end)
+
+	KillAtreides = ordos.AddPrimaryObjective("Kill all Atreides units.")
+	GatherSpice = player.AddPrimaryObjective("Harvest " .. tostring(ToHarvest[Map.Difficulty]) .. " Solaris worth of Spice.")
+	KillOrdos = player.AddSecondaryObjective("Eliminate all Ordos units and reinforcements\nin the area.")
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/atreides03a.lua]: Wow, the objectives of the d2k missions are so varied :p

Does this change at any point in the game?

Comment (abcdefg30) [inline mods/d2k/maps/atreides-03a/atreides03a.lua]: I added the text for the secondary objective(s) myself. Only the primary objective(s)are from the original.

Feel free to suggest another description. :)

File: mods/d2k/maps/atreides-03a/atreides03a-AI.lua (hunk 4 above)
Comment (abcdefg30) [inline mods/d2k/maps/atreides-03a/atreides03a-AI.lua]: I usually add a new line at the top, but okay, removed.

File: mods/d2k/maps/atreides-03a/atreides03a.lua (hunk 9 above)
Comment (abcdefg30) [inline mods/d2k/maps/atreides-03a/atreides03a.lua]: Yep, ops.

**Comment by abcdefg30** [comment]
Should be fixed now.

File: mods/d2k/maps/atreides-03a/atreides03a.lua (hunk 10 above)
Comment (penev92) [inline mods/d2k/maps/atreides-03a/atreides03a.lua]: Yeah Dune 2000 only has primary objectives, and I believe it's one objective per mission.

File: mods/d2k/maps/atreides-03a/map.yaml [hunk 11]
```diff
@@ -0,0 +1,209 @@
+MapFormat: 7
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-03a/map.yaml]: Please rebase on top of bleed so the hex color stuff gets applied.

**Comment by obrakmann** [comment]
Other than that, :+1:

**Comment by abcdefg30** [comment]
Updated.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/0ab655ddf52d33637ed33306250dbce0a5861d49)

**Pull request merged**